Add new hooks actionCategoryProductsTotalBefore and actionCategoryProductsBefore

gwojcik7 opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | **actionCategoryProductsTotalBefore**: Allows modules to modify the SQL query used to get the total number of products in a category. <br>**actionCategoryProductsBefore**: Allows modules to modify the SQL query used to get the products in a category. <br><br>In many customer stores, I had to override the Category::getProducts method to prevent products that are out of stock from being displayed. With these hooks, it will be possible to add a condition for quantity. I also rewrote the SQL to DbQuery so that modifications using Hook could be implemented in a safer way.
| Type?             | improvement
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | 1. Register hooks in the module. <br>2. Make a modification in the query object (e.g., add "where" for products with quantity < 0 that should not be queried).<br>3. You must add this modification in both hooks so that the query from the number of products does not diverge from the query from the products themselves.<br>4. Reset the stock of one of the selected products. <br>5. Check if the product shows on its category.
| UI Tests          | Please run UI tests and paste here the link to the run. As we support MySQL and MariaDB in our tests, you have to launch 2 campaigns (by choosing both). [Visit the UI Tests repo and follow instructions](https://github.com/PrestaShop/ga.tests.ui.pr/).
| Fixed issue or discussion?     |
| Related PRs       | 
| Sponsor company   | Arkonsoft

ps-jarvis commented:
<!-- PR_TABLE_DESCRIPTION_ERROR -->
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `category` should be one of these: `FO`, `BO`, `CO`, `IN`, `WS`, `TE`, `LO`, `ME` or `PM`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

gwojcik7 committed:
Refactor product query handling in Category class and add new hooks

- Replaced raw SQL queries with DbQuery for better readability and maintainability in the Category class.
- Introduced two new hooks: `actionCategoryProductsTotalBefore` and `actionCategoryProductsBefore` to allow modules to modify SQL queries for product totals and product retrieval in a category.
- Enhanced the structure of the SQL query building process to improve performance and clarity.

Hlavtox commented:
@gwojcik7 There are already hooks for this in ProductListingFrontController that allows to modify the product query, result, everything

kpodemski commented:
@Hlavtox 

You're not wrong, but without those hooks, there's no way to modify the query unless you use the faceted search module. I've been in the same situation as @gwojcik7 described, where changing the retrieval of products from the category required overrides.

gwojcik7 committed:
cs fixer

kpodemski commented on classes/Category.php:
```diff
@@ -979,16 +979,40 @@ public function getProducts(
+            }
+
+            Hook::exec('actionCategoryProductsTotalBefore', [
+                'query' => $query,
```
@gwojcik7 You sure you don't need to do `&$query`?

gwojcik7 replied:
@kpodemski Objects are passed by reference by default, so there is no need to add this arbitrarily. I tested it, and works as is.

<img width="473" height="87" alt="image" src="https://github.com/user-attachments/assets/bebae700-a0f1-4e89-a5d1-8fa803384ce3" />

Hlavtox commented on install-dev/data/xml/hook.xml:
```diff
@@ -5478,5 +5478,15 @@
     </hook>
+    <hook id="actionCategoryProductsTotalBefore">
+      <name>actionCategoryProductsTotalBefore</name>
+      <title>Called before getting the total number of products in a category</title>
```
Guys, I think it would be good to specify that this hook will work only if the default provider is used.

Touxten replied:
I aggre; The developers couldn't understand why the hook wasn't working.

I added the label ‘need documentation’ to ensure that the information was added to the documentation.

gwojcik7 committed:
Enhance description for actionCategoryProductsBefore hook to clarify its limitations with custom product providers.

gwojcik7 committed:
Update description for actionCategoryProductsTotalBefore hook to specify its limitation with custom product providers.

kpodemski reviewed (changes requested):
Rebase needed.
